pkg/object/websocketserver/proxy.go

```diff
@@ -0,0 +1,294 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package websocketserver
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/gorilla/websocket"
+
+	"github.com/megaease/easegress/pkg/logger"
+	"github.com/megaease/easegress/pkg/supervisor"
+)
+
+var (
+	// defaultUpgrader specifies the parameters for upgrading an HTTP
+	// connection to a WebSocket connection.
+	defaultUpgrader = &websocket.Upgrader{
+		ReadBufferSize:  1024,
+		WriteBufferSize: 1024,
+	}
+
+	// defaultDialer is a dialer with all fields set to the default zero values.
+	defaultDialer = websocket.DefaultDialer
+
+	// defaultIntervalSecond is the default interval for polling websocket client and server.
+	defaultIntervalSecond time.Duration = 1
+)
+
+// Proxy is an handler that takes an incoming WebSocket
+// connection and proxies it to the backend server.
+type Proxy struct {
+	// server is the HTTPServer
+	server    http.Server
+	superSpec *supervisor.Spec
+
+	// BackendURL URL is the
+	backendURL *url.URL
+
+	// Upgrader specifies the parameters for upgrading a incoming HTTP
+	// connection to a WebSocket connection.
+	upgrader *websocket.Upgrader
+
+	//  Dialer contains options for connecting to the backend WebSocket server.
+	dialer *websocket.Dialer
+
+	done chan struct{}
+}
+
+// NewProxy returns a new Websocket proxy.
+func newProxy(superSpec *supervisor.Spec) *Proxy {
+	proxy := &Proxy{
+		superSpec: superSpec,
+		done:      make(chan struct{}),
+	}
+	go proxy.run()
+	return proxy
+}
+
+// buildReq builds a URL with backend in spec and original HTTP request.
+func (p *Proxy) buildReq(r *http.Request) *url.URL {
+	u := *p.backendURL
+	u.Fragment = r.URL.Fragment
+	u.Path = r.URL.Path
+	u.RawQuery = r.URL.RawQuery
+	return &u
+}
+
+// passMsg passes websocket message from src to dst.
+func (p *Proxy) passMsg(src, dst *websocket.Conn, errc chan error, stop chan struct{}) {
+	handle := func() bool {
+		msgType, msg, err := src.ReadMessage()
+		if err != nil {
+			m := websocket.FormatCloseMessage(websocket.CloseNormalClosure, fmt.Sprintf("%v", err))
+			if e, ok := err.(*websocket.CloseError); ok {
+				if e.Code != websocket.CloseNoStatusReceived {
+					m = websocket.FormatCloseMessage(e.Code, e.Text)
+				}
+			}
+			dst.WriteMessage(websocket.CloseMessage, m)
+			errc <- err
+			return false
+		}
+		err = dst.WriteMessage(msgType, msg)
+		if err != nil {
+			errc <- err
+			return false
+		}
+		return true
+	}
+
+	for {
+		select {
+		// this request handling is stopped due to some error or websocketserver shutdown.
+		case <-stop:
+			return
+		case <-time.After(defaultIntervalSecond * time.Second):
+			if !handle() {
+				return
+			}
+		}
+	}
+}
+
+// run runs the websocket proxy.
+func (p *Proxy) run() {
+	spec := p.superSpec.ObjectSpec().(*Spec)
+	backendURL, err := url.Parse(spec.Backend)
```

Comment:
The `Backend` should be validated in `Spec.Validate` or use `jsonschema:"required,url"